src/tools/tool_custom.rs

```diff
@@ -0,0 +1,225 @@
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::Arc;
+use serde_json::Value;
+use tokio::io::BufReader;
+use async_trait::async_trait;
+use regex::Regex;
+use tokio::sync::{Mutex as AMutex, RwLock as ARwLock};
+use tracing::{info, warn};
+use tokio::process::{Command, Child, ChildStdout, ChildStderr};
+use tokio::time::{timeout, Duration};
+use std::process::Stdio;
+
+use crate::at_commands::at_commands::AtCommandsContext;
+use crate::tools::tools_description::{AtParamDict, Tool};
+use crate::call_validation::{ChatMessage, ChatContent, ContextEnum};
+use crate::global_context::GlobalContext;
+use crate::integrations::process_io_utils::{kill_process_and_children, read_until_token_or_timeout};
+use crate::integrations::sessions::IntegrationSession;
+
+
+pub struct ToolCustom {
+    pub name: String,
+    #[allow(dead_code)]
+    pub parameters: Vec<AtParamDict>,
+    pub parameters_required: Vec<String>,
+    pub command: String,
+    pub runs_in_background: bool,
+    pub runs_in_background_false_timeout: usize,
```

Comment:
blocking: bool
background: bool

Reply:
blocking_timeout: usize
background_wait_keyword: String    -- until something appears in stdout stderr
background_wait_port: u16              -- until port becomes available